File: alc/astnodes.c
// AST node functions for the alic compiler.
// (c) 2025 Warren Toomey, GPL3

#include "alic.h"
#include "proto.h"

// Build and return a generic AST node
ASTnode *mkastnode(int op, ASTnode * left, ASTnode * mid, ASTnode * right) {
  ASTnode *n;

  // Malloc a new ASTnode
  n = (ASTnode *) calloc(1, sizeof(ASTnode));
  if (n == NULL)
    fatal("Unable to calloc in mkastnode()");

  // Copy in the field values and return it
  n->op = op;
  n->left = left;
  n->mid = mid;
  n->right = right;
  return (n);
}

// Make an AST leaf node
ASTnode *mkastleaf(int op, Type * type, bool rvalue,
		   Sym * sym, uint64_t intval) {
  ASTnode *n;
  n = mkastnode(op, NULL, NULL, NULL);
  n->type = type;
  n->rvalue = rvalue;
  n->sym = sym;
  n->litval.intval = intval;
  return (n);
}

// Recursively free an AST tree
void freeAST(ASTnode * n) {
  if (n == NULL)
    return;

  freeAST(n->left);
  freeAST(n->mid);
  freeAST(n->right);

  if (n->strlit != NULL)
    free(n->strlit);
  free(n);
}


// List of AST node names
static char *astname[] = { NULL,
  "ASSIGN", "CAST",
  "ADD", "SUBTRACT", "MULTIPLY", "DIVIDE", "NEGATE",
  "EQ", "NE", "LT", "GT", "LE", "GE", "NOT",
  "AND", "OR", "XOR", "INVERT",
  "LSHIFT", "RSHIFT",
  "NUMLIT", "IDENT", "PRINT", "GLUE", "IF", "WHILE", "FOR",
  "TYPE", "STRLIT", "LOCAL", "FUNCCALL"
};

// Given an AST tree, print it out and follow the
// traversal of the tree that genAST() follows
void dumpAST(ASTnode * n, int level) {
  if (n == NULL)
    fatal("NULL AST node\n");

  // General AST node handling
  for (int i = 0; i < level; i++)
    fprintf(Debugfh, " ");

  if (n->type != NULL) {
    fprintf(Debugfh, "%s ", get_typename(n->type));
  }

  fprintf(Debugfh, "%s ", astname[n->op]);

  switch (n->op) {
  case A_NUMLIT:
    if (n->type->kind >= TY_FLT32)
      fprintf(Debugfh, "%f", n->litval.dblval);
    else
      fprintf(Debugfh, "%ld", n->litval.intval);
    break;
  case A_ASSIGN:
    fprintf(Debugfh, "%s = ", n->sym->name);
    break;
  case A_LOCAL:
    fprintf(Debugfh, "%s", n->sym->name);
    break;
  case A_IDENT:
    if (n->rvalue)
      fprintf(Debugfh, "rval %s", n->sym->name);
    else
      fprintf(Debugfh, "%s", n->sym->name);
    break;
  case A_PRINT:
  case A_FUNCCALL:
    fprintf(Debugfh, "\"%s\"\n", n->left->strlit);
    if (n->right) dumpAST(n->right, level + 2);
    return;
  }

  fprintf(Debugfh, "\n");

  // Reset the level if an A_LOCAL node
  if (n->op == A_LOCAL) level -= 2;

  // General AST node handling
  if (n->left)
    dumpAST(n->left, level + 2);
  if (n->mid)
    dumpAST(n->mid, level + 2);
  if (n->right)
    dumpAST(n->right, level + 2);
}

File: alpy/astnodes.py
import sys
from typing import Optional

from defs import ASTNode, ASTNodeType, TypeKind, fatal, quote_string
from typs import TypeProcessor

_astname = [
    None,  # 0
    "ASSIGN", "CAST",  # 1-2
    "ADD", "SUBTRACT", "MULTIPLY", "DIVIDE", "NEGATE",  # 3-7
    "EQ", "NE", "LT", "GT", "LE", "GE", "NOT",  # 8-14
    "AND", "OR", "XOR", "INVERT",  # 15-18
    "LSHIFT", "RSHIFT",  # 19-20
    "NUMLIT", "IDENT", "PRINT", "GLUE", "IF", "WHILE", "FOR",  # 21-27
    "TYPE", "STRLIT", "LOCAL", "FUNCTION", "FUNCCALL",  # 28-32
    "PRINTF", "MOD", "RETURN", "BLOCK",  # 29-36
]


def mk_ast_node(op: ASTNodeType, left: Optional[ASTNode] = None,
                mid: Optional[ASTNode] = None,
                right: Optional[ASTNode] = None) -> ASTNode:
    node = ASTNode(op)
    node.left = left
    node.mid = mid
    node.right = right
    return node


def mk_ast_leaf(op: ASTNodeType, type_: Optional['Type'] = None,
                rvalue: bool = False, sym: Optional['Sym'] = None,
                intval: int = 0) -> ASTNode:
    node = mk_ast_node(op)
    node.type = type_
    node.rvalue = rvalue
    node.sym = sym
    node.numlit.intval = intval
    return node


def free_ast(node: Optional[ASTNode]) -> None:
    if node is None:
        return
    free_ast(node.left)
    free_ast(node.mid)
    free_ast(node.right)
    # Python has garbage collection, so no need to explicitly free memory


def dump_ast(out, node: Optional[ASTNode], level: int = 0) -> None:
    if node is None:
        fatal("NULL AST node")
        return

    # Print indentation
    if out is None:
        out = sys.stdout
    for _ in range(level):
        out.write(" ")

    # Print type if available
    if node.type is not None:
        type_name = TypeProcessor.get_typename(node.type)
        out.write(f"{type_name} ")

    # Print operation name
    op_name = f"Unknown-op({node.op.value})"
    if node.op.value < len(_astname):
        op_name = _astname[node.op.value]
    out.write(f"{op_name} ")

    # Print node-specific information
    if node.op == ASTNodeType.A_IDENT:
        if node.rvalue:
            out.write(f"rval {node.strlit}")
        else:
            out.write(f"{node.strlit}")
    elif node.op == ASTNodeType.A_STRLIT:
        value = quote_string(node.strlit)
        out.write(f"({value})")
    elif node.op == ASTNodeType.A_NUMLIT:
        if node.type and node.type.kind in (TypeKind.TY_FLT32, TypeKind.TY_FLT64):
            out.write(f"({node.numlit.dblval})")
        else:
            out.write(f"({node.numlit.intval})")
    elif node.op == ASTNodeType.A_ASSIGN:
        out.write(f"{node.left.sym.name} = ")
    elif node.op == ASTNodeType.A_LOCAL:
        out.write(f"{node.sym.name}")
    elif node.op in (ASTNodeType.A_PRINT, ASTNodeType.A_PRINTF, ASTNodeType.A_FUNCCALL):
        if node.left:
            value = quote_string(node.left.strlit)
            out.write(f"\"{value}\"")
        if node.right:
            dump_ast(out, node.right, level + 2)
        return

    out.write("\n")

    # Adjust level for local nodes
    new_level = level + 2 if node.op != ASTNodeType.A_LOCAL else level - 2

    # Recursively dump children
    if node.left:
        dump_ast(out, node.left, new_level)
    if node.mid:
        dump_ast(out, node.mid, new_level)
    if node.right:
        dump_ast(out, node.right, new_level)
